File: sdk/python/bananas/queue.py
"""
Bananas Redis queue implementation.

This module provides the RedisQueue class for job queue operations.
"""

import time
from typing import Optional

import redis

from .exceptions import ConnectionError as BananasConnectionError
from .exceptions import JobNotFoundError
from .models import Job, JobPriority


class RedisQueue:
    """Redis-backed job queue.

    This class handles enqueueing jobs and retrieving job data from Redis.
    It uses the same key patterns and data structures as the Go implementation.

    Attributes:
        client: Redis client instance
        key_prefix: Prefix for all Redis keys (default: "bananas")
    """

    def __init__(self, redis_url: str, key_prefix: str = "bananas"):
        """Initialize Redis queue connection.

        Args:
            redis_url: Redis connection URL (e.g., "redis://localhost:6379/0")
            key_prefix: Prefix for all Redis keys (default: "bananas")

        Raises:
            BananasConnectionError: If connection to Redis fails
        """
        self.key_prefix = key_prefix
        try:
            self.client = redis.from_url(
                redis_url,
                decode_responses=False,  # We handle encoding/decoding ourselves
                socket_connect_timeout=5,
                socket_keepalive=True,
            )
            # Test connection
            self.client.ping()
        except redis.RedisError as e:
            raise BananasConnectionError(f"Failed to connect to Redis: {e}") from e

    def enqueue(self, job: Job) -> None:
        """Enqueue a job to the appropriate priority queue.

        Args:
            job: Job to enqueue

        Raises:
            BananasConnectionError: If Redis operation fails
        """
        try:
            # Store job data
            job_key = f"{self.key_prefix}:job:{job.id}"
            job_json = job.to_json()
            self.client.set(job_key, job_json.encode("utf-8"))

            # Add to appropriate priority queue
            queue_key = f"{self.key_prefix}:queue:{job.priority.value}"
            self.client.lpush(queue_key, job.id.encode("utf-8"))
        except redis.RedisError as e:
            raise BananasConnectionError(f"Failed to enqueue job: {e}") from e

    def enqueue_scheduled(self, job: Job) -> None:
        """Enqueue a job to the scheduled queue.

        Args:
            job: Job to enqueue (must have scheduled_for set)

        Raises:
            BananasConnectionError: If Redis operation fails
            ValueError: If job.scheduled_for is not set
        """
        if job.scheduled_for is None:
            raise ValueError("Job must have scheduled_for set to enqueue to scheduled queue")

        try:
            # Store job data
            job_key = f"{self.key_prefix}:job:{job.id}"
            job_json = job.to_json()
            self.client.set(job_key, job_json.encode("utf-8"))

            # Add to scheduled set with timestamp as score
            scheduled_key = f"{self.key_prefix}:queue:scheduled"
            timestamp = job.scheduled_for.timestamp()
            self.client.zadd(scheduled_key, {job.id.encode("utf-8"): timestamp})
        except redis.RedisError as e:
            raise BananasConnectionError(f"Failed to enqueue scheduled job: {e}") from e

    def get_job(self, job_id: str) -> Optional[Job]:
        """Retrieve a job by its ID.

        Args:
            job_id: Job identifier

        Returns:
            Job instance if found, None otherwise

        Raises:
            BananasConnectionError: If Redis operation fails
        """
        try:
            job_key = f"{self.key_prefix}:job:{job_id}"
            job_json = self.client.get(job_key)

            if job_json is None:
                return None

            return Job.from_json(job_json.decode("utf-8"))
        except redis.RedisError as e:
            raise BananasConnectionError(f"Failed to get job: {e}") from e
        except Exception as e:
            raise JobNotFoundError(f"Failed to deserialize job: {e}") from e

    def close(self) -> None:
        """Close the Redis connection."""
        if self.client:
            self.client.close()

    def __enter__(self) -> "RedisQueue":
        """Context manager entry."""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        """Context manager exit."""
        self.close()

File: sdk/typescript/src/queue.ts
/**
 * Redis queue operations for Bananas.
 */

import Redis from "ioredis";
import { ConnectionError, JobNotFoundError } from "./errors";
import { createJob, jobFromJSON, jobToJSON } from "./models";
import { Job } from "./types";

/**
 * Redis-backed job queue.
 */
export class RedisQueue {
  private redis: Redis;
  private keyPrefix: string;

  /**
   * Create a new Redis queue.
   *
   * @param redisUrl - Redis connection URL
   * @param keyPrefix - Prefix for all Redis keys (default: 'bananas')
   */
  constructor(redisUrl: string, keyPrefix: string = "bananas") {
    this.keyPrefix = keyPrefix;

    try {
      this.redis = new Redis(redisUrl, {
        retryStrategy: (times) => {
          if (times > 3) {
            return null; // Stop retrying
          }
          return Math.min(times * 50, 2000);
        },
      });

      this.redis.on("error", (err) => {
        console.error("Redis connection error:", err);
      });
    } catch (error) {
      throw new ConnectionError(`Failed to connect to Redis: ${error}`);
    }
  }

  /**
   * Enqueue a job to the appropriate priority queue.
   *
   * @param job - Job to enqueue
   */
  async enqueue(job: Job): Promise<void> {
    try {
      const jobKey = `${this.keyPrefix}:job:${job.id}`;
      const queueKey = `${this.keyPrefix}:queue:${job.priority}`;

      // Store job data
      await this.redis.set(jobKey, jobToJSON(job));

      // Add to priority queue
      await this.redis.lpush(queueKey, job.id);
    } catch (error) {
      throw new ConnectionError(`Failed to enqueue job: ${error}`);
    }
  }

  /**
   * Enqueue a job to the scheduled queue.
   *
   * @param job - Job to enqueue (must have scheduledFor set)
   */
  async enqueueScheduled(job: Job): Promise<void> {
    if (!job.scheduledFor) {
      throw new Error("Job must have scheduledFor set");
    }

    try {
      const jobKey = `${this.keyPrefix}:job:${job.id}`;
      const scheduledKey = `${this.keyPrefix}:queue:scheduled`;

      // Store job data
      await this.redis.set(jobKey, jobToJSON(job));

      // Add to scheduled set with timestamp as score
      const timestamp = job.scheduledFor.getTime() / 1000; // Convert to seconds
      await this.redis.zadd(scheduledKey, timestamp, job.id);
    } catch (error) {
      throw new ConnectionError(`Failed to enqueue scheduled job: ${error}`);
    }
  }

  /**
   * Get a job by its ID.
   *
   * @param jobId - Job identifier
   * @returns Job if found, null otherwise
   */
  async getJob(jobId: string): Promise<Job | null> {
    try {
      const jobKey = `${this.keyPrefix}:job:${jobId}`;
      const jobJson = await this.redis.get(jobKey);

      if (!jobJson) {
        return null;
      }

      return jobFromJSON(jobJson);
    } catch (error) {
      if (error instanceof SyntaxError) {
        throw new JobNotFoundError(`Failed to deserialize job: ${error}`);
      }
      throw new ConnectionError(`Failed to get job: ${error}`);
    }
  }

  /**
   * Close the Redis connection.
   */
  async close(): Promise<void> {
    await this.redis.quit();
  }
}
